lib/Transformation.ts

```diff
@@ -1,31 +1,33 @@
 import * as RDF from 'rdf-js';
 import * as RDFString from 'rdf-string';
-import { Algebra as Alg } from 'sparqlalgebrajs';
+import {Algebra as Alg} from 'sparqlalgebrajs';
 
 import * as E from './expressions';
+import {AsyncExtensionApplication, SimpleApplication} from './expressions';
 import * as C from './util/Consts';
+import {TypeURL as DT} from './util/Consts';
 import * as Err from './util/Errors';
+import {ExtensionFunctionError} from './util/Errors';
 import * as P from './util/Parsing';
 
-import {
-  namedFunctions,
-  regularFunctions,
-  specialFunctions,
-} from './functions';
-import { TypeURL as DT } from './util/Consts';
+import {namedFunctions, regularFunctions, specialFunctions,} from './functions';
+import {AsyncExtensionFunction, AsyncExtensionFunctionCallback} from './evaluators/AsyncEvaluator';
+import {SyncExtensionFunction, SyncExtensionFunctionCallback} from './evaluators/SyncEvaluator';
 
-export function transformAlgebra(expr: Alg.Expression): E.Expression {
-  if (!expr) { throw new Err.InvalidExpression(expr); }
+type ExtensionCallbackType<T extends boolean> = T extends true ? SyncExtensionFunctionCallback : AsyncExtensionFunctionCallback;
```

Comment:
I don't mind changing it. I know @rubensworks had a vision for how this should work. @rubensworks what do you think of the proposed signatures?

Reply:
Sure, I'm also fine with something like `ExtensionCallback = {type: 'sync', callback: SyncCallback} | {type: 'async', callback: AsyncCallback}`. No strong opinion on this myself :-)

Reply:
I think instead of calling it `ExtensionCallback` we might want t call it `UserFunctionCreatorConfig` if we start using `AsyncExtensionFunctionCreator´ this might make sense and we can keep this name when adding custom functions.

Reply:
Sure, but perhaps just `FunctionCreatorConfig` then.